File: SimTriggers/src/com/IconPippi/simtriggers/scripting/ScriptLoader.java
package com.IconPippi.simtriggers.scripting;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;

import javax.script.Invocable;
import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import javax.script.ScriptException;

import com.IconPippi.simtriggers.module.Module;
import com.IconPippi.simtriggers.module.ModuleManager;
import com.IconPippi.simtriggers.utils.FileUtils;
import com.IconPippi.simtriggers.utils.Logger;

public class ScriptLoader {
	
	private ModuleManager mm;
	private final Logger logger = new Logger();
	
	public final static ScriptEngineManager engineManager = new ScriptEngineManager();
	public final static ScriptEngine engine = engineManager.getEngineByName("nashorn");
	
	/**
	 * Loads every installed module's scripts
	 */
	public void loadModules() {
		logger.log("Loading scripts...");
		
		mm = new ModuleManager();
		
		final FileUtils fileUtils = new FileUtils();
		
		for (Module m : mm.getModules()) {
			for (File f : fileUtils.getFilesInDir(m.getDir(), false)) {
				
				try {
					if (f.getName().toLowerCase().endsWith(".js") && engine.eval(compileScripts(f)) != null) {
						engine.eval(compileScripts(f));
					}
				} catch (Exception e) {
					logger.log(e.toString());
				}
			}
		}
	}
	
	/**
	 * Execute a function inside JS code
	 * @param Function's name
	 * @param Function's arguments
	 */
	public void invokeFunction(String method, Object... args) {
		final Invocable invoc = (Invocable) engine;
		try {
			invoc.invokeFunction(method, args);
		} catch (NoSuchMethodException | ScriptException e) {
			logger.error(e.toString());
		}
	}
	
	/*
	 * Util method to compile scripts from a .js file
	 */
	private String compileScripts(File f) throws IOException {
		if (f.isDirectory()) return null;
	      
	      StringBuilder compiledScript = new StringBuilder();
	      

	      @SuppressWarnings("resource")
		BufferedReader br = new BufferedReader(new InputStreamReader(new FileInputStream(f), "UTF8"));
	      String line;
	      while ((line = br.readLine()) != null) {
	          compiledScript.append(line).append("\n");
	      }
	      
	      return compiledScript.toString();
	}
	
}

File: src/main/kotlin/com/simtriggers/fsx/scripting/ScriptLoader.kt
package com.simtriggers.fsx.scripting

import javax.script.ScriptEngine
import com.simtriggers.fsx.module.Module
import com.simtriggers.fsx.module.ModulesManager
import com.simtriggers.fsx.SimTriggers
import com.simtriggers.fsx.data.request.RequestData
import com.simtriggers.fsx.event.RegisterEvent
import com.simtriggers.fsx.triggers.TriggerRegister
import com.simtriggers.fsx.triggers.TriggersManager
import com.simtriggers.fsx.util.FileUtils
import java.net.URL
import java.net.URLClassLoader
import dev.iconpippi.logger.Logger
import jdk.nashorn.api.scripting.NashornScriptEngine
import java.io.File
import javax.script.ScriptException
import javax.script.Invocable


/**
 * 02/08/2019
 * Handles the loading of all module's scripts
 *
 * @author IconPippi
 */
class ScriptLoader {

    companion object {
        /** Nashorn JavaScript engine */
        private lateinit var engine: NashornScriptEngine
    }

    // Module manager
    private val mm: ModulesManager = ModulesManager()

    /**
     * Load all modules inside modules' folder
     */
    fun load() {
        // Clear triggers
        TriggerRegister.unregisterAll()

        /** @author ChatTriggers development team (https://github.com/ChatTriggers) */
        val jars = mm.getModules().map {
            it.location.listFiles()?.toList() ?: listOf()
        }.flatten().filter {
            it.name.endsWith(".jar")
        }.map {
            it.toURI().toURL()
        }

        engine = instanceScriptEngine(jars)

        val simTriggersDevKit = FileUtils.saveResource(
            "/simTriggersDevKit.js", File("${mm.modulesFolder}/simTriggersDevKit.js")
        )

        val modulesThread = Thread({
            try {
                engine.eval(simTriggersDevKit)
            } catch (e: java.lang.Exception) {
                Logger.error("An unexpected error occurred while loading simTriggersDevKit.js file.")
                e.printStackTrace()
            }

            mm.getModules().forEach {
                try {
                    engine.eval(compileScripts(it))
                } catch(e: Exception) {
                    Logger.error("Error on loading module: ${it.metadata.moduleName}")
                    e.printStackTrace()
                }
            }
        },"SimTriggers/Modules-Thread")
        modulesThread.stop()
        modulesThread.run()
    }

    /**
     * Execute a function inside JS code
     * @param method Function's name
     * @param args Function's arguments
     */
    fun invokeFunction(method: String?, vararg args: Any?) {
        val invoc = engine as Invocable

        try {
            invoc.invokeFunction(method, *args)
        } catch (e: NoSuchMethodException) {
            Logger.error(e.toString())
        } catch (e: ScriptException) {
            Logger.error(e.toString())
        }

    }

    /**
     * Compile all script files for a module into one single string for evaluation
     * @param m Target module
     */
    @Throws(Exception::class)
    private fun compileScripts(m: Module): String {
        return m.getScriptFiles().joinToString("\n") {
            it.readText()
        }
    }

    /**
     * @author ChatTriggers development team (https://github.com/ChatTriggers)
     */
    @Suppress("DEPRECATION")
    private fun instanceScriptEngine(files: List<URL>): jdk.nashorn.api.scripting.NashornScriptEngine {
        val ucl = URLClassLoader(files.toTypedArray(), SimTriggers::class.java.classLoader)

        return jdk.nashorn.api.scripting.NashornScriptEngineFactory().getScriptEngine(ucl) as jdk.nashorn.api.scripting.NashornScriptEngine
    }

}